React mathml support

louisscruz opened the pull request:
This adds support to React for [MathML](https://developer.mozilla.org/en-US/docs/Web/MathML) as [discussed previously](https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/69884).

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://developer.mozilla.org/en-US/docs/Web/MathML
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

## Notes

I originally opened https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69986 to add support for MathML. In that PR, my initial approach was to supply a minimal set of types, leaving out any marked as legacy, deprecated, or non-standard by [MDN](https://developer.mozilla.org/en-US/docs/Web/MathML). However, in discussion of that PR, it was mentioned that we ought to [prefer types that describe the spec](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69986#issuecomment-2213248221), which seems like a better approach.

The specification for "MathML" is kind of complicated, and in fact there are two variants of specification. MathML and MathML Core from the Math WG:

* [MathML 3 specification](https://www.w3.org/TR/MathML3/)
* [MathML Core specification](https://www.w3.org/TR/mathml-core/)

As MDN puts summarizes:

> MDN uses [MathML Core](https://w3c.github.io/mathml-core/) as a reference specification but, due to an erratic standardization history, legacy MathML features may still show up in existing implementations and web content.

and

> Note: It is highly recommended that developers and authors switch to MathML Core, perhaps relying on other web technologies to cover missing use cases. The Math WG is maintaining a set of [MathML polyfills](https://github.com/mathml-refresh/mathml-polyfills) to facilitate that transition.

The approach taken in this PR is to use the MathML Core specification.

In adjusting the branch to use the MathML Core basis, I fortunately found that very few changes were needed.

* Most notably, the [`maction` element](https://developer.mozilla.org/en-US/docs/Web/MathML/Element/maction) needed to be added.
* Aside from that, this PR is different from the old PR in that it also adds the relevant types to the `ts5.0` directory.

jakebailey commented:
Maybe a silly question, but why is it that these need to be declared in React, if every browser has them? Wouldn't this all have been declared in the DOM types shipped with TypeScript, generated by https://github.com/microsoft/TypeScript-DOM-lib-generator?

(I might just be misunderstanding how this works, given I see other elements defined above the new code too.)

petamoriken commented:
@jakebailey My understanding is that [TypeScript-DOM-lib-generator](https://github.com/microsoft/TypeScript-DOM-lib-generator) describes Web Standards type defs to access the DOM from JavaScript, and @types/react describes type defs to enable JSX to handle it.

The following is a code fragment that makes this relationship easier to understand:

```ts
// TypeScript-DOM-lib-generator

// type defs for using MathML from JavaScript
interface MathMLElement extends Element, ElementCSSInlineStyle, GlobalEventHandlers, HTMLOrSVGElement {
  addEventListener<K extends keyof MathMLElementEventMap>(type: K, listener: (this: MathMLElement, ev: MathMLElementEventMap[K]) => any, options?: boolean | AddEventListenerOptions): void;
  addEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions): void;
  removeEventListener<K extends keyof MathMLElementEventMap>(type: K, listener: (this: MathMLElement, ev: MathMLElementEventMap[K]) => any, options?: boolean | EventListenerOptions): void;
  removeEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | EventListenerOptions): void;
}
```

```ts
// @types/react

// minimum type defs for users do not use TypeScript-DOM-lib-generator types
interface Element {}
interface MathMLElement extends Element {}

// type defs for using MathML from JSX
interface MathMLMActionElement extends MathMLElement {}
interface MathMLMathElement extends MathMLElement {}
// ...

declare module "react" {
  namespace JSX {
    interface IntrinsicElements {
      maction: React.MathMLProps<React.MathMLMActionAttributes, MathMLMActionElement>;
      math: React.MathMLProps<React.MathMLMathAttributes, MathMLMathElement>;
      // ...
    }
  }
}
```

petamoriken commented on types/react/index.d.ts:
```diff
@@ -4186,6 +4335,39 @@ declare namespace React {
             view: React.SVGProps<SVGViewElement>;
+
+            // MathML
+            maction: React.MathMLProps<React.MathMLMActionAttributes, MathMLMActionElement>;
```
https://github.com/w3c/mathml-core/issues/279#issuecomment-2625378665

louisscruz replied:
Added in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/71187/commits/04bb0cc0eda306dd6274bc02c4c366acaa6f9eee

jakebailey commented:
> The following is a code fragment that makes this relationship easier to understand:

Thanks, that does help.

I don't know if I am a great person to approve this otherwise; it would really be good to get the take from a maintainer of the react types.

eps1lon commented on types/react/index.d.ts:
```diff
@@ -3749,6 +3751,153 @@ declare namespace React {
+    interface MathMLAttributes<T> extends DOMAttributes<T> {
+        className?: string | undefined;
+        dir?: "ltr" | "rtl" | undefined;
+        displaystyle?: boolean | undefined;
```
This currently warns for booleans as far as I can tell: https://codesandbox.io/p/sandbox/tender-feistel-ym2mcg?file=%2Fsrc%2Findex.js%3A11%2C23. I think you need to use `"false" | "true"` for now.

louisscruz replied:
Nice catch.

* Types updated in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/71187/commits/ae57a4d5babf168190484685d50d081b77f3647b
* Tests related to this updated in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/71187/commits/c8b502aba0df8d12888a2da49d7bc1a6722b08d8

eps1lon commented:
@louisscruz  This is a pretty big change so I've been slower to respond than usual. Can you address the current comments?

jakebailey commented:
Not yet, @typescript-bot

jakebailey commented:
Not yet (again), @typescript-bot

louisscruz commented on types/react/index.d.ts:
```diff
@@ -3729,6 +3731,153 @@ declare namespace React {
+    interface MathMLMmultiscriptsAttributes extends MathMLAttributes<MathMLMmultiscriptsElement> {}
+    interface MathMLMnAttributes extends MathMLAttributes<MathMLMnElement> {}
+    interface MathMLMoAttributes extends MathMLAttributes<MathMLMoElement> {
+        /* This attribute is non-standard. */
```
Updated throughout in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/71187/commits/ef3b8e7dd897da9c6934e6f57f52a0bfe6aea73e

louisscruz commented:
Just giving another bump here @eps1lon

louisscruz commented:
@Jessidhia Thank you for the thorough review. I've just updated to address the comments:

* I've removed the concept of `MathMLElement` sub-interfaces. Originally I had these in place to mimic the structure of the SVG types and they had just stuck around without much thought. Removing it is quite nicer given the interfaces now match what arbitrary MathML elements' `.__proto__` is at runtime.
* The `*Attributes` interface names have been updated to better align with the SVG pascal case style.

The casing for attribute names is a trickier subject. It looks like there are currently inconsistencies in place. For example, only one property for `webview` is camelCase (`allowFullScreen`) while the remaining properties are not (`allowpopups`, etc.). I do think that keeping the property close to the specification in the case of MathML will probably provide a better developer experience. I'm imagining that coping, pasting, then adjusting will be a common way for interacting with MathML, and it'd be quite tedious to have to update attribute name casing frequently. For now, I've set it up such that:

* `tabIndex` is corrected. I think this makes a lot of sense. (Aside: I wonder if casing differences could be constrained to certain interface levels. Why isn't `tabIndex` defined at the `DOMAttributes` or `AriaAttributes` layer so that such a common element property is consistent across all element types?)
* The other attribute names remain as they are in the specification.
* A comment is added about the strategy used for casing in property names.

I'm happy to adjust, but I thought this might be a reasonable starting place.

louisscruz commented:
Just giving this PR a small nudge. Curious where folks think this stands at this point. @Jessidhia @eps1lon @petamoriken

otomad reviewed (changes requested):
It seems that aria attributes are also available for mathml elememts.

otomad commented on types/react/ts5.0/index.d.ts:
```diff
@@ -3758,6 +3759,159 @@ declare namespace React {
+    interface MathMLMOAttributes extends MathMLAttributes {
+        /** This attribute is non-standard. */
+        accent?: boolean | undefined;
+        fence?: boolean | undefined;
```
`<mo>` element missing `form` attribute.

```suggestion
        fence?: boolean | undefined;
        form?: "prefix" | "infix" | "postfix" | undefined;
```
